coderd/unhanger/detector.go

```diff
@@ -0,0 +1,356 @@
+		defer d.cancel()
+
+		for {
+			select {
+			case <-d.ctx.Done():
+				return
+			case t, ok := <-d.tick:
+				if !ok {
+					return
+				}
+				stats := d.run(t)
+				if stats.Error != nil && !xerrors.As(stats.Error, &acquireLockError{}) {
+					d.log.Warn(d.ctx, "error running workspace build hang detector once", slog.Error(stats.Error))
+				}
+				if len(stats.TerminatedJobIDs) != 0 {
+					d.log.Warn(d.ctx, "detected (and terminated) hung provisioner jobs", slog.F("job_ids", stats.TerminatedJobIDs))
+				}
+				if d.stats != nil {
+					select {
+					case <-d.ctx.Done():
+						return
+					case d.stats <- stats:
+					}
+				}
+			}
+		}
+	}()
+}
+
+// Wait will block until the detector is stopped.
+func (d *Detector) Wait() {
+	<-d.done
+}
+
+// Close will stop the detector.
+func (d *Detector) Close() {
+	d.cancel()
+	<-d.done
+}
+
+func (d *Detector) run(t time.Time) Stats {
+	ctx, cancel := context.WithTimeout(d.ctx, 5*time.Minute)
+	defer cancel()
+
+	stats := Stats{
+		TerminatedJobIDs: []uuid.UUID{},
+		Error:            nil,
+	}
+
+	// Find all provisioner jobs that are currently running but have not
+	// received an update in the last 5 minutes.
+	jobs, err := d.db.GetHungProvisionerJobs(ctx, t.Add(-HungJobDuration))
+	if err != nil {
+		stats.Error = xerrors.Errorf("get hung provisioner jobs: %w", err)
+		return stats
+	}
+
+	// Limit the number of jobs we'll unhang in a single run to avoid
+	// timing out.
+	if len(jobs) > MaxJobsPerRun {
+		// Pick a random subset of the jobs to unhang.
+		rand.Shuffle(len(jobs), func(i, j int) {
+			jobs[i], jobs[j] = jobs[j], jobs[i]
+		})
+		jobs = jobs[:MaxJobsPerRun]
+	}
+
+	// Send a message into the build log for each hung job saying that it
+	// has been detected and will be terminated, then mark the job as
+	// failed.
+	for _, job := range jobs {
+		log := d.log.With(slog.F("job_id", job.ID))
+
+		err := unhangJob(ctx, log, d.db, d.pubsub, job.ID)
+		if err != nil && !(xerrors.As(err, &acquireLockError{}) || xerrors.As(err, &jobNotRunningError{})) {
+			log.Error(ctx, "error forcefully terminating hung provisioner job", slog.Error(err))
+			continue
+		}
+
+		stats.TerminatedJobIDs = append(stats.TerminatedJobIDs, job.ID)
+	}
+
+	return stats
+}
+
+func unhangJob(ctx context.Context, log slog.Logger, db database.Store, pub pubsub.Pubsub, jobID uuid.UUID) error {
+	var lowestLogID int64
+
+	err := db.InTx(func(db database.Store) error {
+		locked, err := db.TryAcquireLock(ctx, database.GenLockID(fmt.Sprintf("hang-detector:%s", jobID)))
+		if err != nil {
+			return xerrors.Errorf("acquire lock: %w", err)
+		}
+		if !locked {
+			// This error is ignored.
+			return acquireLockError{}
+		}
+
+		// Refetch the job while we hold the lock.
+		job, err := db.GetProvisionerJobByID(ctx, jobID)
+		if err != nil {
+			return xerrors.Errorf("get provisioner job: %w", err)
+		}
+		jobStatus := db2sdk.ProvisionerJobStatus(job)
+		if jobStatus != codersdk.ProvisionerJobRunning {
```

Comment:
Perhaps overly cautious, but should we check some last update timestamp too so that there's no chance something came in between listing hang jobs and acquiring the lock?

Reply:
Good call, will do!